rust_src/src/windows.rs

```diff
@@ -1536,4 +1592,131 @@ pub fn window_body_width_lisp(window: LispWindowLiveOrSelected, pixelwise: bool)
+/// should be used for buffers that mostly display text from right to left.
+///
+/// If LEFT is non-nil and INVERSE is nil, this means that the y-pixel value
+/// returned for a specific line specifies the distance in pixels from the
+/// left edge of the last (leftmost) glyph of that line to the right edge
+/// (body edge if BODY is non-nil) of WINDOW.  If LEFT and INVERSE are both
+/// non-nil, the y-pixel value returned for a specific line specifies the
+/// distance in pixels from the left edge (body edge if BODY is non-nil) of
+/// WINDOW to the left edge of the last (leftmost) glyph of that line.
+///
+/// Normally, the value of this function is not available while Emacs is
+/// busy, for example, when processing a command.  It should be retrievable
+/// though when run from an idle timer with a delay of zero seconds.
+#[lisp_fn]
+pub fn window_lines_pixel_dimensions(
+    window: LispWindowLiveOrSelected,
+    first: Option<LispNumber>,
+    last: Option<LispNumber>,
+    body: bool,
+    inverse: bool,
+    left: bool,
+) -> LispObject {
+    let window: LispWindowRef = window.into();
+    let buffer: LispBufferRef = window.contents.into();
+    let max_y = if body {
+        window.text_bottom_y()
+    } else {
+        window.pixel_height
+    };
+    let window_width = if body {
+        window.body_width(true)
+    } else {
+        window.pixel_width
+    };
+    let header_line_height = window.header_line_height();
+    let subtract = if body { header_line_height } else { 0 };
+    let matrix = window.get_matrix();
+
+    if unsafe { noninteractive } || window.is_pseudo() {
+        return Qnil;
+    }
+
+    // Fail if current matrix is not up to date.
+    unsafe {
+        if !window.window_end_valid()
+            || windows_or_buffers_changed > 0
+            || buffer.clip_changed()
+            || buffer.prevent_redisplay_optimizations_p()
+            || window.is_outdated()
+        {
+            return Qnil;
+        }
+    }
+
+    let mut row = if let Some(first) = first {
+        matrix.row(first.to_fixnum() as usize)
+    } else {
+        if body {
+            matrix.first_text_row()
+        } else {
+            matrix.row(0)
+        }
+    };
+
+    let end_row = if let Some(last) = last {
+        matrix.row(last.to_fixnum() as usize)
+    } else {
+        if body {
+            matrix.bottom_text_row(window)
+        } else {
+            unsafe { matrix.row_unchecked(matrix.nrows as usize) }
+        }
+    };
+
+    let mut rows = Qnil;
+    while row.as_ptr() <= end_row.as_ptr() && row.enabled_p() && row.y + row.height < max_y {
+        if left {
+            let glyph = unsafe { &*row.glyphs[TEXT_AREA as usize] };
+            let width = if inverse {
+                glyph.pixel_width as i32
+            } else {
+                window_width - glyph.pixel_width as i32
+            };
+            rows = (LispObject::cons(width, row.y + row.height - subtract), rows).into();
```

Comment:
The last line of the if and else are the same. Can this be moved to be common after the if/else?

I think you can write `rows = ((width, row.y + row.height - subtract), rows).into();`. No?

Reply:
Yep, you can. My mistake when trying to do something like this was calling `.into()` on the inner tuple.